fix fr reset api

tushar00jain opened the pull request:
Summary:
- there are various places that access fr's `entries_` field
- if we empty the entries_ on reset, the accesses can result in an error
- so we only perform a soft delete instead of clearing out the entries copletely
  - only reset id_ on the reset
  - keep track of a reset_epoch which increments everytime reset is called
  - dump_entries only returns entries from the latest epoch
  - api's that access entries also check if the reset epoch matches
- make the `next_` always track the index in the circular buffer - this change was needed to make the soft delete's implementation easier

---
[//]: # (BEGIN SAPLING FOOTER)
Stack created with [Sapling](https://sapling-scm.com). Best reviewed with [ReviewStack](https://reviewstack.dev/pytorch/pytorch/pull/165642).
* #166103
* #166102
* __->__ #165642

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci

fduwjj commented on torch/csrc/distributed/c10d/FlightRecorderDetail.hpp:
```diff
@@ -178,32 +187,57 @@ std::vector<typename FlightRecorder<EventType>::Entry> FlightRecorder<
+  auto it = reset_epoch_start_idx_.find(reset_epoch);
+  TORCH_CHECK(it != reset_epoch_start_idx_.end());
+  // Calculate idx based on where the epoch started
+  return (it->second + id) % max_entries_;
```
I think this is wrong? because you reset id to 0, we should not consider reset_epoch anymore? I thought reset_epoch is only for dump and retire_id. When you try to get entry, you can just directly get that entry using the index?

fduwjj commented:
Thanks for making FR working for TorchFT. Can we mention the context and rationale on why we are doing this in the PR description? Like because we have restart and we want to reset FR without race condition so that we want to do soft-reset?

Also I have questions on the correctness of that getting entry. Maybe you can clarify a bit more? Also you can write a unit test (we already have one in test_c10d_nccl) to validate your change as well.

tushar00jain commented:
@fduwjj 

> like because we have restart and we want to reset FR without race condition so that we want to do soft-reset?

this is the first 2 bullet points?

> Also you can write a unit test (we already have one in test_c10d_nccl) to validate your change as well.

added

tushar00jain commented on torch/csrc/distributed/c10d/FlightRecorderDetail.hpp:
```diff
@@ -214,8 +248,8 @@ void FlightRecorder<EventType>::retire_id(
-  Entry* entry = &entries_.at(*id % max_entries_);
-  if (entry->id_ == *id) {
+  Entry* entry = &entries_.at(getIdxFromId(*id, *reset_epoch));
+  if (entry->id_ == *id && entry->reset_epoch_ == *reset_epoch) {
```
yep i think better to keep it a noop inside flight recorder so we don't have to implement the same logic in all pg's

fduwjj commented on test/distributed/test_c10d_nccl.py:
```diff
@@ -4626,6 +4626,229 @@ def test_fr_record_reset(self, timing_enabled):
+            return
+
+        # Override buffer size to 10 for faster testing
+        os.environ["TORCH_NCCL_TRACE_BUFFER_SIZE"] = "10"
```
nit: unset this ENV in the end so that it does not interfere with other unit tests.

fduwjj reviewed (approved):
I think this will work, just some nit in torch/csrc/distributed/c10d/FlightRecorderDetail.hpp. Thanks!

tushar00jain commented on torch/csrc/distributed/c10d/FlightRecorderDetail.hpp:
```diff
@@ -104,15 +110,20 @@ std::optional<size_t> FlightRecorder<EventType>::record(
+
+  if (next_ == max_entries_) {
+    next_ = 0;
+  }
```
this is needed, the pr changes how `next_` is tracked. we use it to track the index in the buffer at all times. we do this so we can create the mapping from reset epoch to the index in the buffer

tushar00jain committed:
fix fr reset api

Summary:
- there are various places that access fr's `entries_` field
- if we empty the entries_ on reset, the accesses can result in an error
- so we only perform a soft delete instead of clearing out the entries copletely
  - only reset id_ on the reset
  - keep track of a reset_epoch which increments everytime reset is called
  - dump_entries only returns entries from the latest epoch
  - api's that access entries also check if the reset epoch matches
- make the `next_` always track the index in the circular buffer - this change was needed to make the soft delete's implementation easier
